lib/vrl/compiler/src/vm/machine.rs

```diff
@@ -0,0 +1,533 @@
+    /// each one.
+    /// The VM is stack based. When the `Return` OpCode is encountered the top item on the stack is popped and returned.
+    /// It is expected that the final instruction is a `Return`.
+    pub fn interpret<'a>(&self, ctx: &mut Context<'a>) -> Result<Value, ExpressionError> {
+        // Any mutable state during the run is stored here.
+        let mut state: VmState = VmState::new(self);
+
+        loop {
+            let next = state.next()?;
+
+            match next {
+                OpCode::Abort => {
+                    // Aborts the process.
+                    let start = state.next_primitive()?;
+                    let end = state.next_primitive()?;
+                    return Err(ExpressionError::Abort {
+                        span: Span::new(start, end),
+                    });
+                }
+                OpCode::Return => {
+                    // Ends the process and returns the top item from the stack - or Null if the stack is empty.
+                    return Ok(state.stack.pop().unwrap_or(Value::Null));
+                }
+                OpCode::Constant => {
+                    let value = state.read_constant()?;
+                    state.stack.push(value);
+                }
+                OpCode::Not => match state.stack.pop() {
+                    None => return Err("Notting nothing".into()),
+                    Some(Value::Boolean(value)) => state.stack.push(Value::Boolean(!value)),
+                    _ => return Err("Notting non boolean".into()),
+                },
+                OpCode::Add => binary_op(&mut state, Value::try_add)?,
+                OpCode::Subtract => binary_op(&mut state, Value::try_sub)?,
+                OpCode::Multiply => binary_op(&mut state, Value::try_mul)?,
+                OpCode::Divide => binary_op(&mut state, Value::try_div)?,
+                OpCode::Rem => binary_op(&mut state, Value::try_rem)?,
+                OpCode::Merge => binary_op(&mut state, Value::try_merge)?,
+                OpCode::Greater => binary_op(&mut state, Value::try_gt)?,
+                OpCode::GreaterEqual => binary_op(&mut state, Value::try_ge)?,
+                OpCode::Less => binary_op(&mut state, Value::try_lt)?,
+                OpCode::LessEqual => binary_op(&mut state, Value::try_le)?,
+                OpCode::NotEqual => {
+                    let rhs = state.pop_stack()?;
+                    let lhs = state.pop_stack()?;
+                    state.stack.push((!lhs.eq_lossy(&rhs)).into());
+                }
+                OpCode::Equal => {
+                    let rhs = state.pop_stack()?;
+                    let lhs = state.pop_stack()?;
+                    state.stack.push(lhs.eq_lossy(&rhs).into());
+                }
+                OpCode::Pop => {
+                    // Removes the top item from the stack
+                    let _ = state.stack.pop();
```

Comment:
Why does this ignore errors? Wouldn't this only fail if there was a compilation error (which _should_ return an error)?

Reply:
Good catch, it shouldn't!